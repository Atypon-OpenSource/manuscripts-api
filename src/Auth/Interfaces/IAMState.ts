/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Represents data passed in "state" param for IAM Auth flows
 */
export interface IAMState {
  [key: string]: string | null
  /**
   * Device ID passed by the client
   */
  deviceId: string,
  /**
   * Uri tracked by the client application to return the user back to the originally requested url, before
   * OAuth flow was started by the client application
   */
  redirectUri: string | null,
  /**
   * Represents UI theme used by the client application
   */
  theme: string | null
}
