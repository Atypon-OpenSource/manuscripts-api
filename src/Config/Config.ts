/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as path from 'path'
import { load } from 'dotenv-safe'
import { isString, isNumber } from '../util'
import { CreateBucketOptions } from 'couchbase'
import {
  clientApplicationsFromSplitString,
  scopeConfigurationsFromSplitString,
} from '../Models/ClientApplicationModels'
import { existsSync, readFileSync } from 'fs'
import { log } from '../Utilities/Logger'
import { ConfigurationError } from '../Errors'
import {
  APIConfiguration,
  DatabaseConfiguration,
  AuthConfiguration,
  AWSConfiguration,
  GoogleConfiguration,
  EmailConfiguration,
  GatewayConfiguration,
  ServerConfiguration,
  ClientApplicationsConfiguration,
  EnvironmentLike,
  Environment,
  ConfigurationContainer,
  IAMConfiguration,
  ScopedAccessTokenConfiguration,
  LiteratumConfiguration,
  PressroomConfiguration,
  TemplateConfiguration,
  ShacklesConfiguration,
} from './ConfigurationTypes'
import { normalizeURL } from './normalize-url'

const isUTF8 = require('is-utf8')

/**
 * Configuration string validator:
 * passes input parameter through as return value
 * if it is a non-empty string, throws otherwise.
 */
function getString(value: any, key: string): string {
  if (!isString(value) || value.length === 0) {
    throw new ConfigurationError(key, value)
  }
  return value
}

function getArray(value: string, key: string, separator: string): string[] {
  if (value === '') {
    return []
  }
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

function getOptionalArray(envVar: string | undefined, key: string, separator: string): string[] {
  if (envVar === undefined || envVar === '') {
    return []
  }
  const value = getString(envVar, key)
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

// undefined intentionally here,
// in order to avoid creating an object with null values
// (which the relevant 3rd party code may not interpret similarly).
function getStringOptional(value: any): string | undefined {
  if (!isString(value) || value.length === 0) {
    return undefined
  }
  return value
}

/**
 * Configuration number validator:
 * passes input parameter through as return value
 * if it is a number, return null otherwise.
 *
 * The use of undefined (instead of null) is intentional, in order to avoid
 * passing objects with null values to 3rd party code (which may not interpret them correctly).
 */
function getNumberOptional(value: any): number | undefined {
  const numberVal = Number(value)
  if (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal)) {
    return undefined
  }
  return numberVal
}

function getNumber(value: any, key: string, allowMissing?: boolean): number {
  const numberVal = Number(value)
  if (!allowMissing && (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal))) {
    throw new ConfigurationError(key, value)
  }
  return numberVal
}

export class Configuration implements ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly AWS: AWSConfiguration
  readonly google: GoogleConfiguration
  readonly IAM: IAMConfiguration
  readonly email: EmailConfiguration
  readonly gateway: GatewayConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly literatum: LiteratumConfiguration
  readonly pressroom: PressroomConfiguration
  readonly shackles: ShacklesConfiguration
  readonly template: TemplateConfiguration

  constructor(env: EnvironmentLike) {
    const bucketOptions = this.createBucketOptions(env)

    this.API = {
      port: Number(env.APP_PORT),
      oauthStateEncryptionKey: getString(
        env.APP_OAUTH_STATE_ENCRYPTION_KEY,
        'APP_OAUTH_STATE_ENCRYPTION_KEY'
      ),
      hostname: getString(env.APP_HOSTNAME_PUBLIC, 'APP_HOSTNAME_PUBLIC'),
    }

    /* istanbul ignore next */
    const skipVerification =
      env.NODE_ENV === 'development' &&
      Boolean(getNumber(env.APP_SKIP_ACCOUNT_VERIFICATION, 'APP_SKIP_ACCOUNT_VERIFICATION'))

    this.auth = {
      jwtSecret: getString(env.APP_JWT_SECRET, 'APP_JWT_SECRET'),
      skipVerification: skipVerification,
      hashSaltRounds:
        getNumber(env.APP_HASH_SALT_ROUNDS, 'APP_HASH_SALT_ROUNDS', true) ||
        (env.NODE_ENV === Environment.Production ? 10 : 3),
      serverSecret: getString(env.APP_SERVER_SECRET, 'APP_SERVER_SECRET'),
      enableNonConnectAuth: Boolean(
        getNumber(env.APP_ENABLE_NON_CONNECT_AUTH, 'APP_ENABLE_NON_CONNECT_AUTH')
      ),
    }

    const buckets = {
      user: getString(env.APP_USER_BUCKET, 'APP_USER_BUCKET'),
      data: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      project: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      state: getString(env.APP_STATE_BUCKET, 'APP_STATE_BUCKET'),
      derivedData: getString(env.APP_DERIVED_DATA_BUCKET, 'APP_DERIVED_DATA_BUCKET'),
    }

    // database initialization should not be happening without INITIALIZE_DATABASE='true',
    // except for when running tests when it should occur also when INITIALIZE_DATABASE was omitted.
    /* istanbul ignore next */
    const initialize =
      env.INITIALIZE_DATABASE === 'true' ||
      (typeof env.INITIALIZE_DATABASE === 'undefined' && env.NODE_ENV === Environment.Test)

    this.DB = {
      bucketOptions,
      buckets,
      initializeContents: initialize,
      uri: normalizeURL(getString(env.APP_DB_URI, 'APP_DB_URI')),
      username: getString(env.APP_COUCHBASE_ADMIN_USER, 'APP_COUCHBASE_ADMIN_USER'),
      password: getString(env.APP_COUCHBASE_ADMIN_PASS, 'APP_COUCHBASE_ADMIN_PASS'),
      bucketAdminPassword: getString(
        env.APP_COUCHBASE_RBAC_PASSWORD,
        'APP_COUCHBASE_RBAC_PASSWORD'
      ),
    }

    this.AWS = {
      accessKeyId: getString(env.APP_AWS_ACCESS_KEY_ID, 'APP_AWS_ACCESS_KEY_ID'),
      secretAccessKey: getString(env.APP_AWS_SECRET_ACCESS_KEY, 'APP_AWS_SECRET_ACCESS_KEY'),
      region: getString(env.APP_AWS_REGION, 'APP_AWS_REGION'),
    }

    this.google = {
      clientID: getString(env.APP_GOOGLE_CLIENT_ID, 'APP_GOOGLE_CLIENT_ID'),
      clientSecret: getString(env.APP_GOOGLE_CLIENT_SECRET, 'APP_GOOGLE_CLIENT_SECRET'),
      authCallback: normalizeURL(
        getString(env.APP_GOOGLE_AUTH_CALLBACK, 'APP_GOOGLE_AUTH_CALLBACK')
      ),
    }

    this.IAM = {
      clientID: getString(env.APP_IAM_CLIENT_ID, 'APP_IAM_CLIENT_ID'),
      authServerURL: getString(env.APP_IAM_SERVER_URL, 'APP_IAM_SERVER_URL'),
      authServerPermittedURLs: Array.from(
        new Set(
          getArray(
            getString(env.APP_IAM_PERMITTED_SERVER_URLS, 'APP_IAM_PERMITTED_SERVER_URLS'),
            'APP_IAM_PERMITTED_SERVER_URLS',
            ';'
          )
        )
      ),
      authCallbackPath: normalizeURL(
        getString(env.APP_IAM_AUTH_CALLBACK_PATH, 'APP_IAM_AUTH_CALLBACK_PATH')
      ),
      libraryURL: normalizeURL(getString(env.APP_IAM_LIBRARY_URL, 'APP_IAM_LIBRARY_URL')),
      apiServerURL: Array.from(
        new Set(
          getArray(
            getString(env.APP_API_SERVER_URL, 'APP_API_SERVER_URL'),
            'APP_API_SERVER_URL',
            ';'
          )
        )
      ),
    }

    this.email = {
      fromAddress: getString(env.APP_FROM_EMAIL, 'APP_FROM_EMAIL'),
      fromBaseURL: normalizeURL(getString(env.APP_BASE_URL, 'APP_BASE_URL')),
    }

    this.gateway = {
      cookieDomain: getString(env.APP_GATEWAY_COOKIE_DOMAIN, 'APP_GATEWAY_COOKIE_DOMAIN'),
      hostname: getString(env.APP_GATEWAY_HOSTNAME, 'APP_GATEWAY_HOSTNAME'),
      ports: {
        admin: getString(env.APP_GATEWAY_ADMIN_PORT, 'APP_GATEWAY_ADMIN_PORT'),
        public: getString(env.APP_GATEWAY_PUBLIC_PORT, 'APP_GATEWAY_PUBLIC_PORT'),
      },
    }

    const host = this.email.fromBaseURL.replace(/https{0,1}\:\/\//, '')
    const additionalOrigins = [`http://${host}`, `https://${host}`]
    this.server = {
      storeOnlySSLTransmittedCookies: Boolean(
        getNumber(
          env.APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES,
          'APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES'
        )
      ),
      allowedCORSOrigins: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_CORS_ORIGINS, 'APP_ALLOWED_CORS_ORIGINS'),
            'APP_ALLOWED_CORS_ORIGINS',
            ';'
          ).concat(additionalOrigins)
        )
      ), // get unique values from potentially duplicated ones.
    }

    this.literatum = {
      allowedIPAddresses: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_IP_ADDRESSES, 'APP_ALLOWED_IP_ADDRESSES'),
            'APP_ALLOWED_IP_ADDRESSES',
            ';'
          )
        )
      ),
    }

    this.apps = {
      knownClientApplications: clientApplicationsFromSplitString(
        getString(env.APP_CLIENT_APPLICATIONS, 'APP_CLIENT_APPLICATIONS'),
        ';',
        ','
      ),
    }

    this.scopes = scopeConfigurationsFromSplitString(
      getString(env.APP_CONTAINER_TOKEN_SCOPES, 'APP_CONTAINER_TOKEN_SCOPES'),
      ';',
      ','
    )

    this.pressroom = {
      baseurl: getString(env.APP_PRESSROOM_BASE_URL, 'APP_PRESSROOM_BASE_URL'),
      apiKey: getString(env.APP_PRESSROOM_APIKEY, 'APP_PRESSROOM_APIKEY'),
    }

    this.shackles = {
      baseUrl: getString(env.SHACKLES_URL, 'SHACKLES_URL'),
    }

    this.template = {
      allowedOwners: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_OWNERS,
        'APP_COUCHBASE_ALLOWED_OWNERS',
        ';'
      ),
      allowedProjects: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_PROJECTS,
        'APP_COUCHBASE_ALLOWED_PROJECTS',
        ';'
      ),
    }
  }

  static fromEnv(envExamplePath: string) {
    if (existsSync('.env') && isUTF8(readFileSync('.env'))) {
      const dotenvResult = load({
        allowEmptyValues: false,
        sample: envExamplePath,
      })

      // hack to ignore the case of missing .env
      if (dotenvResult && dotenvResult.error) {
        throw dotenvResult.error
      }
    } else {
      log.error('.env is missing or is not a UTF8 encoded text file.')
    }

    return process.env // load() call above loads .env file into process.env
  }

  public createBucketOptions(env: NodeJS.ProcessEnv): CreateBucketOptions {
    const authType = getStringOptional(env.APP_DB_AUTH_TYPE)
    const bucketType = getStringOptional(env.APP_DB_BUCKET_TYPE)
    const ramQuotaMB = getNumberOptional(env.APP_DB_RAM_QUOTA_MB)
    const replicaNumber = getNumberOptional(env.APP_DB_REPLICA_NUMBER)

    const bucketOptions: CreateBucketOptions = {}
    if (authType) {
      bucketOptions.authType = authType
    }
    if (bucketType) {
      bucketOptions.bucketType = bucketType
    }
    if (ramQuotaMB) {
      bucketOptions.ramQuotaMB = ramQuotaMB
    }
    if (replicaNumber) {
      bucketOptions.replicaNumber = replicaNumber
    }

    return bucketOptions
  }
}

const env = Configuration.fromEnv(path.join(__dirname, `../../.env`))
export const config = new Configuration(env)
