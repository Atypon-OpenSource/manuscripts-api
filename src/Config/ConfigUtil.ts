/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ValidationError } from '../Errors'

/**
 * Returns a map of key/value pairs from the passed config string.
 * Expects each key/value pair to be separated by `outerSeparator` and the "key" and "value" to be separated by
 * `innerSeparator`. For eg. for input string - "a1:b1;a2:b2" with outerSeparator as ";" and innerSeparator as ":",
 * it will return a map containing "a1" and "a2" as keys and "b1" and "b2" as corresponding values.
 * @param configStr - input string
 * @param innerSeparator - marker that separates the key/value pairs
 * @param outerSeparator - marker that separates key and the value
 */
export function getMap (configStr: string, innerSeparator: string, outerSeparator: string): Map<string, string> {
  let map = new Map()
  configStr.split(outerSeparator).map(x => {
    const configParts = x.split(innerSeparator)
    if (configParts.length !== 2) {
      throw new ValidationError('Expected two values in the config', configParts)
    }
    map.set(configParts[0], configParts[1])
  })
  return map
}
