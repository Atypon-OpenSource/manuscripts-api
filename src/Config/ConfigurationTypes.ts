/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RSA_JWK } from 'pem-jwk'

import { ClientApplication } from '../Models/ClientApplicationModels'

export enum Environment {
  Test = 'test',
  Development = 'development',
  Production = 'production',
}

/**
 * Represents an API configuration.
 */
export type APIConfiguration = {
  readonly port: number
  readonly hostname: string
}

/**
 * Represents a database configuration.
 */
export enum BucketKey {
  User = 'user',
  Project = 'project',
}

export type DatabaseConfiguration = {
  readonly buckets: { [name in BucketKey]: string }
}

export type ScopedAccessTokenConfiguration = {
  readonly name: string
  readonly secret: string
  readonly publicKeyPEM: string | null
  readonly publicKeyJWK: RSA_JWK | null
  readonly expiry: number
  readonly identifier: string // maps to the "kid" property when represented in JWK.
}

/**
 * Represents authentication configuration.
 */
export type AuthConfiguration = {
  readonly jwtSecret: string
  readonly skipVerification: boolean
  readonly hashSaltRounds: number
  readonly serverSecret: string
  readonly enableNonConnectAuth: boolean
}

/**
 * Represents email sending configuration.
 */
export type EmailConfiguration = {
  readonly fromAddress: string
  readonly fromBaseURL: string
}

export type ClientApplicationsConfiguration = {
  readonly knownClientApplications: Array<ClientApplication>
}

export type ServerConfiguration = {
  readonly allowedCORSOrigins: ReadonlyArray<string>
}

export interface EnvironmentLike {
  [key: string]: string | undefined
}

export interface PressroomConfiguration {
  readonly baseurl: string
  readonly apiKey: string
}

export interface ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly email: EmailConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly pressroom: PressroomConfiguration
}
