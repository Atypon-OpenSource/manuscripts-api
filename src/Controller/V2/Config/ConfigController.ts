/*!
 * Â© 2023 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DIContainer } from '../../../DIContainer/DIContainer'
import { BaseController } from '../../BaseController'

const localesPath = __dirname + '/../../../../dist/config/csl/locales/'
const sharedPath = __dirname + '/../../../../dist/config/shared/'
const stylesPath = __dirname + '/../../../../dist/config/csl/styles/'

export class ConfigController extends BaseController {
  private async getData(path: string, fileName?: string, ids?: string[] | string) {
    if (fileName) {
      fileName = fileName.endsWith('.json') ? fileName : fileName + '.json'
      return await DIContainer.sharedContainer.configService.loadConfigData(path, fileName, ids)
    }
    return await DIContainer.sharedContainer.configService.loadConfigData(path)
  }

  async getSharedData(fileName?: string, ids?: any) {
    return this.getData(sharedPath, fileName, ids)
  }

  async getLocales(fileName?: string) {
    return this.getData(localesPath, fileName)
  }

  async getStyles(fileName?: string) {
    return this.getData(stylesPath, fileName)
  }
}
