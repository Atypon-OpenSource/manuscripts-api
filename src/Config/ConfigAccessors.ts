/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { config } from './Config'
import { BucketKey } from './ConfigurationTypes'
import { IllegalStateError } from '../Errors'

// The following functions are free functions in a separate module from Config to make mocking independently of the Configuration object easier.

export function appDataPublicGatewayBaseURI (): string {
  return `http://${config.gateway.hostname}:${config.gateway.ports.public}`
}

export function appDataAdminGatewayBaseURI (): string {
  return `http://${config.gateway.hostname}:${config.gateway.ports.admin}`
}

export function appDataPublicGatewayURI (bucketKey: BucketKey): string {
  const bucket = config.DB.buckets[bucketKey]
  if (!bucket) {
    throw new IllegalStateError(`No bucket with key '${bucketKey}' amongst configured buckets with public gateway URI: ${JSON.stringify(config.DB.buckets)}`, config.DB.buckets)
  }
  return `${appDataPublicGatewayBaseURI()}/${bucket}`
}

export function appDataAdminGatewayURI (bucketKey: BucketKey): string {
  const bucket = config.DB.buckets[bucketKey]
  if (!bucket) {
    throw new IllegalStateError(`No bucket with key '${bucketKey}' amongst configured buckets with admin gateway URI: ${JSON.stringify(config.DB.buckets)}`, config.DB.buckets)
  }
  return `${appDataAdminGatewayBaseURI()}/${bucket}`
}
