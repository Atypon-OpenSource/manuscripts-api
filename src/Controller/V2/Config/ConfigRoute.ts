/*!
 * Â© 2023 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { celebrate } from 'celebrate'
import { NextFunction, Request, Response, Router } from 'express'
import { StatusCodes } from 'http-status-codes'

import { AuthStrategy } from '../../../Auth/Passport/AuthStrategy'
import { BaseRoute } from '../../BaseRoute'
import { ConfigController } from './ConfigController'
import { defaultSchema, sharedSchema } from './ConfigSchema'

export class ConfigRoute extends BaseRoute {
  private get basePath(): string {
    return '/config'
  }
  private configController = new ConfigController()

  public create(router: Router): void {
    router.get(
      `${this.basePath}/shared/:fileName?/:id?`,
      celebrate(sharedSchema),
      AuthStrategy.JsonHeadersValidation,
      AuthStrategy.JWTAuth,
      (req: Request, res: Response, next: NextFunction) => {
        return this.runWithErrorHandling(async () => {
          await this.getSharedData(req, res)
        }, next)
      }
    )
    router.get(
      `${this.basePath}/locales/:fileName?`,
      celebrate(defaultSchema),
      AuthStrategy.JsonHeadersValidation,
      AuthStrategy.JWTAuth,
      (req: Request, res: Response, next: NextFunction) => {
        return this.runWithErrorHandling(async () => {
          await this.getLocales(req, res)
        }, next)
      }
    )
    router.get(
      `${this.basePath}/styles/:fileName?`,
      celebrate(defaultSchema),
      AuthStrategy.JsonHeadersValidation,
      AuthStrategy.JWTAuth,
      (req: Request, res: Response, next: NextFunction) => {
        return this.runWithErrorHandling(async () => {
          await this.getStyles(req, res)
        }, next)
      }
    )
  }
  private async getSharedData(req: Request, res: Response) {
    let data
    const fileName = req.params.fileName
    const id = req.params.id
    if (fileName) {
      data = await this.configController.getSharedData(fileName, id)
    } else {
      data = await this.configController.getSharedData()
    }
    data ? res.status(StatusCodes.OK).send(data) : res.status(StatusCodes.NO_CONTENT)
  }
  private async getLocales(req: Request, res: Response) {
    let data
    const fileName = req.params.fileName
    if (fileName) {
      data = await this.configController.getLocales(fileName)
    } else {
      data = await this.configController.getLocales('locales.json')
    }
    data ? res.status(StatusCodes.OK).send(data) : res.status(StatusCodes.NO_CONTENT)
  }
  private async getStyles(req: Request, res: Response) {
    let data
    const fileName = req.params.fileName
    if (fileName) {
      data = await this.configController.getStyles(fileName)
    } else {
      data = await this.configController.getStyles()
    }
    data ? res.status(StatusCodes.OK).send(data) : res.status(StatusCodes.NO_CONTENT)
  }
}
