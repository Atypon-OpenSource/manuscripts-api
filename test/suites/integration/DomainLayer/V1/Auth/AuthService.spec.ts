/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { drop, seed, testDatabase, dropBucket } from '../../../../../utilities/db'
import {
  InvalidCredentialsError,
  AccountNotFoundError
} from '../../../../../../src/Errors'
import { TEST_TIMEOUT } from '../../../../../utilities/testSetup'
import {
  validGoogleAccess,
  validGoogleAccessWithInvitationId
} from '../../../../../data/fixtures/googleAccess'
import { DIContainer } from '../../../../../../src/DIContainer/DIContainer'
import { SeedOptions } from '../../../../../../src/DataAccess/Interfaces/SeedOptions'
import { BucketKey } from '../../../../../../src/Config/ConfigurationTypes'

jest.setTimeout(TEST_TIMEOUT)

let db: any = null
beforeAll(async () => (db = await testDatabase()))
afterAll(() => db.bucket.disconnect())

describe('AuthService login', () => {
  const seedOptions: SeedOptions = { users: true, applications: true }

  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })

  test('should fail if email is invalid', () => {
    const authService = DIContainer.sharedContainer.authService

    return expect(
      authService.login({
        email: 'invalid-email@example.com',
        password: '123456',
        deviceId: 'deviceId',
        appId: 'appId'
      })
    ).rejects.toThrowError(AccountNotFoundError)
  })

  test('should fail if password is invalid', () => {
    const authService = DIContainer.sharedContainer.authService

    return expect(
      authService.login({
        email: 'valid-user@manuscriptsapp.com',
        password: 'invalid pass',
        deviceId: 'deviceId',
        appId: 'appId'
      })
    ).rejects.toThrowError(InvalidCredentialsError)
  })

  test('should return authorized user', async () => {
    const authService = DIContainer.sharedContainer.authService

    const authorizedUser = await authService.login({
      email: 'valid-user@manuscriptsapp.com',
      password: '12345',
      deviceId: 'deviceId',
      appId: 'appId'
    })

    expect(authorizedUser.token).toBeDefined()
    expect(authorizedUser.user).toBeDefined()
    expect(await DIContainer.sharedContainer.syncService
                      .gatewayAccountExists(authorizedUser.user._id,
                                            BucketKey.DerivedData)).toEqual(true)

    delete (authorizedUser as any).token
    delete (authorizedUser as any).user._id
    delete (authorizedUser as any).syncSessions

    expect(authorizedUser).toMatchSnapshot()

    const anotherLoginAttempt = await authService.login({
      email: 'valid-user@manuscriptsapp.com',
      password: '12345',
      deviceId: 'deviceId',
      appId: 'appId'
    })

    expect(anotherLoginAttempt.token).toBeDefined()
    expect(anotherLoginAttempt.user).toBeDefined()
    expect(await DIContainer
            .sharedContainer
            .syncService
            .gatewayAccountExists(anotherLoginAttempt.user._id,
                                  BucketKey.DerivedData)).toEqual(true)
  })
})

describe('AuthService loginGoogle', () => {
  const seedOptions: SeedOptions = { users: true, invitations: true }
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })

  test('should return authorized user', async () => {
    const authService = DIContainer.sharedContainer.authService
    const authorizedUser = await authService.loginGoogle(validGoogleAccess)
    const authorizedUser2 = await authService.loginGoogle(
      validGoogleAccessWithInvitationId
    )

    expect(authorizedUser.token).toBeDefined()
    expect(authorizedUser.user).toBeDefined()

    delete (authorizedUser as any).token
    delete (authorizedUser as any).user._id
    delete (authorizedUser as any).syncSessions

    expect(authorizedUser).toMatchSnapshot()

    expect(authorizedUser2.token).toBeDefined()
    expect(authorizedUser2.user).toBeDefined()

    delete (authorizedUser2 as any).token
    delete (authorizedUser2 as any).user._id
    delete (authorizedUser2 as any).syncSessions

    expect(authorizedUser2).toMatchSnapshot()
  })

  test('should return authorized user and update existing user token', async () => {
    const authService = DIContainer.sharedContainer.authService
    const userTokenRepository = DIContainer.sharedContainer.userTokenRepository

    let authorizedUser = await authService.loginGoogle(validGoogleAccess)

    const userToken = await userTokenRepository.getById(
      authorizedUser.user._id
    )

    // log user again
    authorizedUser = await authService.loginGoogle(validGoogleAccess)

    // get user token again
    const updatedUserToken = await userTokenRepository.getById(
      authorizedUser.user._id
    )

    expect(userToken).toEqual(updatedUserToken)
  })
})

describe('AuthService refreshSyncSessions', () => {
  const seedOptions: SeedOptions = { users: true }

  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })

  test('should fail if token is invalid', () => {
    const authService = DIContainer.sharedContainer.authService

    return expect(
      authService.refreshSyncSessions('asdasdas')
    ).rejects.toThrowError(InvalidCredentialsError)
  })
})

describe('AuthService IAM', () => {
  const seedOptions: SeedOptions = { users: true }

  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })

  test('OAuth callback should fail with no email in the scope', () => {
    const authService = DIContainer.sharedContainer.authService
    return expect(authService.iamOAuthCallback({ } as any, authService.decodeIAMState('derp'))).rejects.toThrowError(InvalidCredentialsError)
  })

  test('OAuth callback should fail with an invalid payload', () => {
    const authService = DIContainer.sharedContainer.authService
    // base64(x=y:u=v:deviceId=derp) = eD15OnU9djpkZXZpY2VJZD1kZXJwCg==
    return expect(authService.iamOAuthCallback({ email: 'valid-user@manuscriptsapp.com' } as any, authService.decodeIAMState('eD15OnU9djpkZXZpY2VJZD1kZXJwCg=='))).rejects.toThrowError()
  })

  test('IAM OAuth state parsing', () => {
    const authService = DIContainer.sharedContainer.authService
    // base64(x=y:u=v:deviceId=derp) = eD15OnU9djpkZXZpY2VJZD1kZXJwCg==
    return expect(authService.decodeIAMState('eD15OnU9djpkZXZpY2VJZD1kZXJwCg==').deviceId).toEqual('derp\n')
  })
})
