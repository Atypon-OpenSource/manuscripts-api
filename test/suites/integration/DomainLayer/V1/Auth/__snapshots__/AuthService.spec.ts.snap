// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService login should return authorized user 1`] = `
Object {
  "user": Object {
    "email": "valid-user@manuscriptsapp.com",
    "name": "Valid System User",
  },
}
`;

exports[`AuthService loginGoogle should return authorized user 1`] = `
Object {
  "user": Object {
    "connectUserID": undefined,
    "email": "valid-google-user@manuscriptsapp.com",
    "name": "Example User",
  },
}
`;

exports[`AuthService loginGoogle should return authorized user 2`] = `
Object {
  "user": Object {
    "connectUserID": undefined,
    "email": "valid-google-user@manuscriptsapp.com",
    "name": "Example User",
  },
}
`;
